src/Middleware/TrustedHostsNetworkResolver.php

```diff
@@ -0,0 +1,445 @@
+        if ($type === null) {
+            $ipList = $this->getFormattedIpList($ipList);
+        } elseif ($type === self::IP_HEADER_TYPE_RFC7239) {
+            $ipList = $this->getForwardedElements($ipList);
+        }
+        array_unshift($ipList, ['ip' => $actualHost]);  // server's ip to first position
+        $ipDataList = [];
+        do {
+            $ipData = array_shift($ipList);
+            if (!isset($ipData['ip'])) {
+                $ipData = $this->reverseObfuscate($ipData, $ipDataList);
+                if (!isset($ipData['ip'])) {
+                    break;
+                }
+            }
+            $ip = $ipData['ip'];
+            if (!$this->isValidHost($ip, ['any'], $ipValidator)) {
+                break;
+            }
+            $ipDataList[] = $ipData;
+            if (!$this->isValidHost($ip, $trustedHostData[self::DATA_KEY_HOSTS], $ipValidator)) {
+                break;
+            }
+        } while (count($ipList) > 0);
+
+        if ($this->attributeIps !== null) {
+            $request = $request->withAttribute($this->attributeIps, $ipDataList);
+        }
+
+        $uri = $request->getUri();
+        if (isset($ipData['httpHost'])) {
+            $uri = $uri->withHost($ipData['httpHost']);
+        } else {
+            // find host from headers
+            $host = $this->getHttpHost($request, $trustedHostData[self::DATA_KEY_HOST_HEADERS]);
+            if ($host !== null) {
+                $uri = $uri->withHost($host);
+            }
+        }
+        if (isset($ipData['protocol'])) {
+            $uri = $uri->withScheme($ipData['protocol']);
+        } else {
+            // find scheme from headers
+            $scheme = $this->getScheme($request, $trustedHostData[self::DATA_KEY_PROTOCOL_HEADERS]);
+            if ($scheme !== null) {
+                $uri = $uri->withScheme($scheme);
+            }
+        }
+        $urlParts = $this->getUrl($request, $trustedHostData[self::DATA_KEY_URL_HEADERS]);
+        if ($urlParts !== null) {
+            [$path, $query] = $urlParts;
+            $uri = $uri->withPath($path);
+            if ($query !== null) {
+                $uri = $uri->withQuery($query);
+            }
+        }
+        return $handler->handle($request->withUri($uri)->withAttribute('requestClientIp', $ipData['ip']));
+    }
+
+    /**
+     * Validate host by range
+     *
+     * This method can be extendable by overwriting eg. with reverse DNS verification.
+     */
+    protected function isValidHost(string $host, array $ranges, Ip $validator): bool
+    {
+        return $validator->ranges($ranges)->validate($host)->isValid();
+    }
+
+    /**
+     * Reverse obfuscating host data
+     *
+     * The base operation does not perform any transformation on the data.
+     * This method can be extendable by overwriting eg.
+     */
+    protected function reverseObfuscate(array $ipData, array $ipDataList): array
```

Comment:
`$ipDataList` isn't used. Is it here intentionally?

Reply:
Method is protected but it's not overwritten anywhere. Is it meant to be extended?

Reply:
Yes, this is a special case where you can extends with custom function to recover the obfuscated host data. Also receive a `$ipDataList` variable to restore, which contains the valid elements of the chain. It is unknown to me if in such a case these input parameters are sufficient, I am even thinking about passing the `$request` variable.

Alternatively, it can be replaced with a callback.
```php
(new ...NetworkResolver(...))->withReverseObfuscate(function(...) {...})
```

However, this solution is more problematic at enterprise level:
```php
(new ...NetworkResolver(...))->withReverseObfuscate(function(...) {...})
vs
new MyNetworkResolver()
```

Reply:
What's the use case for extending the class?

Reply:
The `Forwarded` header supports obfuscated hosts, eg.
```
                            trusted           trusted             trusted           trusted
                               |                  |                 |                  |
           CLIENT <--------> PROXY <----------> PROXY2 <-------> PROXY3 <---------> SERVER
           1.1.1.1           2.2.2.2             --              3.3.3.3            4.4.4.4
Forwarded:                   for=1.1.1.1,    for=2.2.2.2,     for=_hidden 
```

PROXY2 is present in the network but for some reason is not public.

```php
class MyNetwrokResolver extends Trusted... {
  protected function reverseObfuscate(array $ipData, array $ipDataList): array {
     if($ipData['for'] === '_hidden' && /*before 2.2.2.2*/) {
       $ipData['ip'] = '2.2.2.0';  // any trusted IP
     }
     return parent::reverseObfuscate($ipData, $ipDataList)
  }
}
```

Reply:
This part is still under construction because it is not a good way to overwrite it with a fake IP address.

Reply:
The final of inputs:
```php
    protected function reverseObfuscate(
        array $ipData,
        array $ipDataListValidated,
        array $ipDataListRemaining,
        RequestInterface $request
    ): array {

```